### PR TITLE
Optimize stuck appeals query

### PR DESCRIPTION
**Pull request by pkarman**
The existing query was timing out, even after several minutes.

Rewrite should improve on the existing logic and perform in < 1 second (tested in production).

### CODE REVIEW

File: app/queries/appeals_with_no_tasks_or_all_tasks_on_hold_query.rb [hunk 1]
```diff
@@ -29,33 +29,26 @@ def dispatched_appeals_on_hold
...
     tasks_for(klass_name).where(type: [BvaDispatchTask.name], status: completed)
   end
 
-  def established_appeals
-    Appeal.where.not(established_at: nil)
-  end
-
   def appeals_with_zero_tasks
-    established_appeals.left_outer_joins(:tasks).where(tasks: { id: nil })
+    Appeal.established.left_outer_joins(:tasks).where(tasks: { id: nil })
```
Comment (pkarman) [inline app/queries/appeals_with_no_tasks_or_all_tasks_on_hold_query.rb:37]: `.established` is now a class scope.

File: app/queries/appeals_with_no_tasks_or_all_tasks_on_hold_query.rb [hunk 2]
```diff
@@ -29,33 +29,26 @@ def dispatched_appeals_on_hold
...
   end
 
   def tasks_for(klass_name)
     Task.select(:appeal_id).where(appeal_type: klass_name)
   end
 
-  def stuck_query(klass_name)
-    klass = klass_name.constantize
```
Comment (pkarman) [inline app/queries/appeals_with_no_tasks_or_all_tasks_on_hold_query.rb:53]: abstraction not needed. we run it on AMA appeals only.

File: app/queries/appeals_with_no_tasks_or_all_tasks_on_hold_query.rb [hunk 3]
```diff
@@ -29,33 +29,26 @@ def dispatched_appeals_on_hold
...
-    klass.where.not(id: tasks_for(klass_name).closed)
-      .where.not(id: tasks_for(klass_name).where(type: RootTask.name, status: cancelled))
+  def stuck_appeals
+    Appeal.established.active
       .joins(:tasks)
-      .group("#{table}.id")
-      .having("count(tasks) = count(case when tasks.status = 'on_hold' then 1 end)")
+      .group("appeals.id")
+      .having(
+        "count(case when tasks.status in (?) then 1 end) = count(case when tasks.status = ? then 1 end)",
```
Comment (pkarman) [inline app/queries/appeals_with_no_tasks_or_all_tasks_on_hold_query.rb:49]: old logic was testing that all tasks were on hold. we only care if all open tasks are on hold.

File: spec/queries/appeals_with_no_tasks_or_all_tasks_on_hold_query_spec.rb [hunk 4]
```diff
@@ -8,6 +8,7 @@
     hearing_task = create(:hearing_task, parent: appeal.root_task)
     schedule_hearing_task = create(:schedule_hearing_task, parent: hearing_task)
     appeal.root_task.descendants.each(&:on_hold!)
+    schedule_hearing_task.completed!
```
Comment (pkarman) [inline spec/queries/appeals_with_no_tasks_or_all_tasks_on_hold_query_spec.rb:11]: close some of the tasks but not all.

File: app/queries/appeals_with_no_tasks_or_all_tasks_on_hold_query.rb [hunk 5]
```diff
@@ -29,33 +29,26 @@ def dispatched_appeals_on_hold
...
-    klass = klass_name.constantize
-    table = klass.table_name
-    klass.where.not(id: tasks_for(klass_name).closed)
-      .where.not(id: tasks_for(klass_name).where(type: RootTask.name, status: cancelled))
+  def stuck_appeals
+    Appeal.established.active
       .joins(:tasks)
-      .group("#{table}.id")
-      .having("count(tasks) = count(case when tasks.status = 'on_hold' then 1 end)")
+      .group("appeals.id")
```
Comment (yoomlam) [inline app/queries/appeals_with_no_tasks_or_all_tasks_on_hold_query.rb:47]: `active` scope already `.joins(:tasks).group("appeals.id")`:
https://github.com/department-of-veterans-affairs/caseflow/blob/e6199cc8409da075fe46e55a7ba7bac1bc793663/app/models/appeal.rb#L45

Comment (pkarman) [reply app/queries/appeals_with_no_tasks_or_all_tasks_on_hold_query.rb:47]: @yoomlam that is true, but from an OO encapsulation pov, I don't know that we can count on that always being true.

**Pull request merged**